src/tuner/algo/allreduce/ring.h

```diff
@@ -0,0 +1,311 @@
+/*
+ * Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All rights reserved.
+ */
+
+#ifndef NCCL_OFI_TUNER_ALLREDUCE_RING_H_
+#define NCCL_OFI_TUNER_ALLREDUCE_RING_H_
+
+#include "config.h"
+#include <assert.h>
+#include <math.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "nccl-headers/nvidia/tuner.h"
+
+/**
+ * @brief Number of chunk steps for NCCL_PROTO_LL protocol.
+ */
+#define NCCL_OFI_TUNER_RING_LL_CHUNK_STEPS (1ULL)
+
+/**
+ * @brief Number of chunk steps for NCCL_PROTO_LL128 protocol.
+ */
+#define NCCL_OFI_TUNER_RING_LL128_CHUNK_STEPS (1ULL)
+
+/**
+ * @brief Number of chunk steps for NCCL_PROTO_SIMPLE protocol.
+ */
+#define NCCL_OFI_TUNER_RING_SIMPLE_CHUNK_STEPS (4ULL)
+
+/**
+ * @brief Overhead numerator for NCCL_PROTO_SIMPLE protocol.
+ */
+#define NCCL_OFI_TUNER_RING_SIMPLE_OVERHEAD_NUMERATOR (1ULL)
+
+/**
+ * @brief Overhead denominator for NCCL_PROTO_SIMPLE protocol.
+ */
+#define NCCL_OFI_TUNER_RING_SIMPLE_OVERHEAD_DENOMINATOR (1ULL)
+
+/**
+ * @brief Overhead numerator for NCCL_PROTO_LL128 protocol.
+ */
+#define NCCL_OFI_TUNER_RING_LL128_OVERHEAD_NUMERATOR (15ULL)
+
+/**
+ * @brief Overhead denominator for NCCL_PROTO_LL128 protocol.
+ */
+#define NCCL_OFI_TUNER_RING_LL128_OVERHEAD_DENOMINATOR (16ULL)
+
+/**
+ * @brief Overhead numerator for NCCL_PROTO_LL protocol.
+ */
+#define NCCL_OFI_TUNER_RING_LL_OVERHEAD_NUMERATOR (1ULL)
+
+/**
+ * @brief Overhead denominator for NCCL_PROTO_LL protocol.
+ */
+#define NCCL_OFI_TUNER_RING_LL_OVERHEAD_DENOMINATOR (2ULL)
+
+/**
+ * @brief Number of steps for NCCL protocols.
+ */
+#define NCCL_OFI_TUNER_NCCL_STEPS (8ULL)
+
+/**
+ * @brief Size of NCCL_LL_FIFOLINE structure.
+ */
+#define NCCL_OFI_TUNER_NCCL_SIZEOF_NCCL_LL_FIFOLINE (16ULL)
+
+/**
+ * @brief Size of a warp in CUDA.
+ */
+#define NCCL_OFI_TUNER_NCCL_WARP_SIZE (32ULL)
+
+/**
+ * @brief Maximum number of channels in NCCL.
+ */
+#define NCCL_OFI_TUNER_NCCL_MAXCHANNELS (32ULL)
+
+/**
+ * @brief Maximum number of threads for NCCL protocols.
+ */
+#define NCCL_OFI_TUNER_NCCL_MAX_NTHREADS (640ULL)
+
+/**
+ * @brief Maximum number of threads for NCCL_PROTO_SIMPLE protocol.
+ */
+#define NCCL_OFI_TUNER_NCCL_SIMPLE_MAX_NTHREADS (512ULL)
+
+/**
+ * @brief Maximum number of threads for NCCL_PROTO_LL protocol.
+ */
+#define NCCL_OFI_TUNER_NCCL_LL_MAX_NTHREADS (512ULL)
+
+/**
+ * @brief Number of lines per thread for NCCL_PROTO_LL protocol.
+ */
+#define NCCL_OFI_TUNER_NCCL_LL_LINES_PER_THREAD (8ULL)
+
+/**
+ * @brief Maximum number of threads for NCCL_PROTO_LL128 protocol.
+ */
+#define NCCL_OFI_TUNER_NCCL_LL128_MAX_NTHREADS (640ULL)
+
+/**
+ * @brief Number of elements per thread for NCCL_PROTO_LL128 protocol.
+ */
+#define NCCL_OFI_TUNER_NCCL_LL128_ELEMS_PER_THREAD (120ULL)
+
+/**
+ * @brief Expected data type size.
+ */
+#define NCCL_OFI_TUNER_EXPECTED_DTYPE_SIZE (4ULL)
+
+/**
+ * @brief Buffer size for NCCL_PROTO_SIMPLE protocol.
+ */
+#define NCCL_OFI_TUNER_NCCL_BUFFSIZE (1 << 22)
```

Comment:
how can we override this value if NCCL_BUFFSIZE environment variable is set by user?

Reply:
the tuner should refuse to operate and fallback to the NCCL internal tuner if getenv("NCCL_BUFFSIZE") is not null.

Reply:
why pick up NCCL_BUFFSIZE if it's not NULL?

Reply:
Implemented and amended, something we always planned to do but just wanted to get this out the door.